### ISSUE TITLE
class:list broken

### ISSUE THREAD

**Issue by y-nk** [withastro/astro#7747]
### What version of `astro` are you using?

2.6.1 and above

### Are you using an SSR adapter? If so, which one?

no

### What package manager are you using?

yarn

### What operating system are you using?

macOS

### What browser are you using?

all of them

### Describe the Bug

when passing to an Astro component, `class` and `class:list` are passed as regular props instead of the magical class merging feature similar to `clsx`.

### What's the expected result?

`class` and `class:list` should work together as the famous `clsx` utility as advertised on the doc site.

### Link to Minimal Reproducible Example

https://codesandbox.io/p/sandbox/youthful-lamarr-l9l375?file=/package-lock.json:1256,6-1256,24

### Participation

- [X] I am willing to submit a pull request for this issue **if somebody knowledgeable would give some pointers where to start looking**

**Comment by FineWolf** [comment]
I'm not so sure that's an issue or that auto-magically merging on custom components is really desirable, as, on custom components, `class` could really mean anything (it's not guaranteed that the developer of a component intended that `class` means CSS class; there's no obligation that it MUST mean that).

You can simply pass on the `class` and `class:list` props to the actual child HTML element.

```tsx
---
import type { AstroBuiltinAttributes } from 'astro';
interface Props {
  class?: string | undefined;
  'class:list'?: AstroBuiltinAttributes['class:list']
}

const { class: cssClass, 'class:list': cssClassList } = Astro.props;
---
<div class:list={['always-applied-class-here', cssClass, cssClassList]}>
  <slot />
</div>
```

Some custom components would want the actual raw values of `class:list` for their particular use-case.

That said, it could be an optional opt-in behavior to automatically merge those, but I don't think it should be the default.

**Comment by natemoo-re** [comment]
Hmm, this seems like a possible regression? I'm pretty sure we had tests for this. `class` and `class:list` should merge and `class:list` should be normalized and transformed to a plain `class` before being passed down to components.

**Comment by Hoolean** [comment]
Hello! I am having an issue where a template string in `class` is not interpolated (instead rendered as plain text) when `class:list` is also provided. Could this be related? :)

**Comment by Calvin-LL** [comment]
After spending the past two days looking into this. I think it's time to merge `class:list` into `class`. See my RFC here: https://github.com/withastro/roadmap/discussions/629

**Comment by y-nk** [comment]
definitely nice. i'd love to see that. it's closer to Vue's `class` prop which is one of the most convenient implementation i've experienced in component libraries.

**Comment by natemoo-re** [comment]
I was able to look into this! It was definitely a regression when we refactored some rendering code. The original change split astro and framework component rendering into two different functions, and `class:list` merging into `class` was only handled for framework components.

Unfortunately, our test suite didn't catch this because the [`astro-class-list`](https://github.com/withastro/astro/tree/main/packages/astro/test/fixtures/astro-class-list) fixture didn't catch the user-facing API on `Astro.props`, only the end result (spreading `Astro.props` onto a `span`, where `class:list` was properly handled again.)

I just merged https://github.com/withastro/astro/pull/8142 into the `next` branch, so this will go out in 3.0 so as to not break existing Astro projects that have coded around the current behavior.

As for @Calvin-LL's proposal, thank you for opening [this discussion](https://github.com/withastro/roadmap/discussions/629)! We can keep chatting about it there. We didn't want to rush into making a decision just to get it into 3.0.
